build_resources.py

```diff
@@ -0,0 +1,203 @@
+
+    def _save_to_cache(self, dir):
+        cdir = self._get_cachedir_for(dir)
+        print("Saving build result of %s" % dir)
+        files = self._get_result_files(dir)
+        self._copy_files(files, src=dir, dest=cdir, message="Caching to")
+
+    def _get_cached_build_result(self, dir):
+        result = self._get_cachedir_for(dir)
+        if os.path.isdir(result):
+            return result
+        return None
+
+    def _update_from_cached_result(self, dir, cached_result):
+        print("Using cached result from %s" % cached_result)
+        files = self._get_result_files(dir)
+        self._copy_files(files, src=cached_result, dest=dir,
+                         message="Updating from cache")
+
+    def _copy_files(self, files, src, dest, message=''):
+        for filepath in files:
+            srcpath = os.path.join(src, filepath)
+            destpath = os.path.join(dest, filepath)
+            if message:
+                print('{}: {}'.format(message, destpath))
+            if os.path.exists(srcpath):
+                destdir = os.path.dirname(destpath)
+                if not os.path.isdir(destdir):
+                    os.makedirs(destdir)
+                shutil.copy(srcpath, destpath)
+            elif os.path.exists(destpath):
+                os.remove(destpath)
+
+    def _get_cachedir_for(self, dir):
+        cachedir = self._cachedirs.get(dir)
+        if not cachedir:
+            tag = 'prod' if self.opts.production else 'dev'
+            hash = self._get_dir_hash(dir)
+            cachedir = os.path.join(self.cache_base, tag, hash[:2], hash[2:])
+            self._cachedirs[dir] = cachedir
+        return cachedir
+
+    def _get_dir_hash(self, dir):
+        ignores = set(
+            os.path.join(dir, ignored_item)
+            for ignored_item in self._get_result_files(dir))
+
+        def ignorer(fullpath):
+            basename = os.path.basename(fullpath)
+            return (is_excluded_filename(basename) or fullpath in ignores)
+
+        return hash_path_recursively(dir, ignorer)['checksum'].hexdigest()
+
+    def _get_result_files(self, dir):
+        result_files = self._result_files.get(dir)
+        if result_files is None:
+            result_files = []
+            list_file_path = os.path.join(dir, 'generated_resources.txt')
+            with open(list_file_path, 'rt') as fp:
+                for line in fp:
+                    stripped_line = line.strip()
+                    if stripped_line and not stripped_line.startswith('#'):
+                        file_path = stripped_line.replace('/', os.path.sep)
+                        result_files.append(file_path)
+            self._result_files[dir] = result_files
+        return result_files
+
+
+def hash_path_recursively(path, ignorer=None, hasher=hashlib.sha1):
+    checksum = hasher()
+    size = 0
+    if os.path.isdir(path):
+        tp = 'dir'
+        checksum.update(b'DIR:\n')
+        for item in os.listdir(path):
```

Comment:
I pushed a fixup commit for that. Don't merge before I've rebased the fixup into the original commit.